File: machine/ts/machine/devices/input/output_port.ts
import { IO_INP_OUTPUT_CTRL } from '../../bus/io';
import type { Value } from '../../cpu/cpu';
import { Memory } from '../../memory/memory';
import { InputControllerRegisterFile } from './registers';
import {
	decodeInputOutputIntensityQ16,
	INP_OUTPUT_CTRL_APPLY,
	type InputControllerInputSource,
} from './contracts';

export class InputControllerOutputPort {
	public constructor(
		private readonly input: InputControllerInputSource,
		private readonly registers: InputControllerRegisterFile,
		private readonly memory: Memory,
	) {}

	public static writeOutputControlRegisterThunk(context: InputControllerOutputPort, _addr: number, value: Value): void {
		const command = (value as number) >>> 0;
		if (command === INP_OUTPUT_CTRL_APPLY) {
			context.input.applyInputControllerVibrationEffect(
				context.registers.selectedPadIndex(),
				context.registers.state.outputDurationMs,
				decodeInputOutputIntensityQ16(context.registers.state.outputIntensityQ16),
			);
		}
		context.memory.writeIoValue(IO_INP_OUTPUT_CTRL, 0);
	}
}

File: machine/cpp/machine/devices/input/output_port.cpp
#include "machine/devices/input/output_port.h"

#include "spec/bmsx/io.h"
#include "machine/devices/input/contracts.h"
#include "machine/devices/input/registers.h"
#include "machine/memory/memory.h"

namespace bmsx {

InputControllerOutputPort::InputControllerOutputPort(InputControllerInputSource& input, const InputControllerRegisterFile& registers, Memory& memory)
	: m_input(input)
	, m_registers(registers)
	, m_memory(memory) {
}

void InputControllerOutputPort::writeOutputControlRegisterThunk([[maybe_unused]] u32 addr, u32 value) {
	if (value == INP_OUTPUT_CTRL_APPLY) {
		m_input.applyInputControllerVibrationEffect(
			m_registers.selectedPadIndex(),
			static_cast<f64>(m_registers.state.outputDurationMs),
			decodeInputOutputIntensityQ16(m_registers.state.outputIntensityQ16)
		);
	}
	m_memory.writeIoU32(IO_INP_OUTPUT_CTRL, 0u);
}

} // namespace bmsx
